apis/r/tests/testthat/test-shape.R

```diff
@@ -1,107 +1,222 @@
 
-    if ("int_column" %in% index_column_names) {
-      int_dom <- dom[["int_column"]]
-      int_mxd <- mxd[["int_column"]]
-      expect_true(int_dom[[1]] < -2000000000)
-      expect_true(int_dom[[2]] > 2000000000)
-    }
+  for (i in seq_along(index_column_name_choices)) {
+    index_column_names <- index_column_name_choices[[i]]
 
-    if ("string_column" %in% index_column_names) {
-      expect_equal(dom[["string_column"]], c("", ""))
-      expect_equal(mxd[["string_column"]], c("", ""))
+    for (use_domain_at_create in c(FALSE, TRUE)) {
+
+      uri <- withr::local_tempdir("soma-dataframe-shape")
+
+      # Create
+      if (dir.exists(uri)) unlink(uri, recursive=TRUE)
+
+      domain_for_create <- NULL
+      if (use_domain_at_create) {
+        domain_for_create <- domain_at_create_choices[[i]]
+      }
+
+      sdf <- SOMADataFrameCreate(
+        uri,
+        asch,
+        index_column_names = index_column_names,
+        domain = domain_for_create)
+
+      expect_true(sdf$exists())
+      expect_true(dir.exists(uri))
+
+      # Write
+      tbl0 <- arrow::arrow_table(int_column = 1L:4L,
+                                 soma_joinid = 1L:4L,
+                                 float_column = 1.1:4.1,
+                                 string_column = c("apple", "ball", "cat", "dog"),
+                                 schema = asch)
+
+      sdf$write(tbl0)
+      sdf$close()
+
+      sdf <- SOMADataFrameOpen(uri)
+
+      # Check shape and maxshape et al.
+      if (!.new_shape_feature_flag_is_enabled()) {
+        expect_false(sdf$tiledbsoma_has_upgraded_domain())
+      } else {
+        expect_true(sdf$tiledbsoma_has_upgraded_domain())
+      }
+      expect_error(sdf$shape(), class = "notYetImplementedError")
+      expect_error(sdf$maxshape(), class = "notYetImplementedError")
+
+      # Not implemented this way per
+      # https://github.com/single-cell-data/TileDB-SOMA/pull/2953#discussion_r1746125089
+      # sjid_shape <- sdf$.maybe_soma_joinid_shape()
+      # sjid_maxshape <- sdf$.maybe_soma_joinid_maxshape()
+      soma_context <- soma_context()
+      sjid_shape <- maybe_soma_joinid_shape(sdf$uri, soma_context)
+      sjid_maxshape <- maybe_soma_joinid_maxshape(sdf$uri, soma_context)
+
+      if ("soma_joinid" %in% index_column_names) {
+        # More testing to come on
+        # https://github.com/single-cell-data/TileDB-SOMA/issues/2407
+        expect_false(is.na(sjid_shape))
+        expect_false(is.na(sjid_maxshape))
+      } else {
+        expect_true(is.na(sjid_shape))
+        expect_true(is.na(sjid_maxshape))
+      }
+
+      # Check has_upgraded_domain
+      if (!.new_shape_feature_flag_is_enabled()) {
+        expect_false(sdf$tiledbsoma_has_upgraded_domain())
+      } else {
+        expect_true(sdf$tiledbsoma_has_upgraded_domain())
+      }
+
+      # Check domain and maxdomain
+      dom = sdf$domain()
+      mxd = sdf$maxdomain()
```

Comment:
Consitency in assignment operator
```suggestion
      dom <- sdf$domain()
      mxd <- sdf$maxdomain()
```